sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/optimizer/expressions.scala

```diff
@@ -212,27 +212,27 @@ object ReorderAssociativeOperator extends Rule[LogicalPlan] {
  * 1. Converts the predicate to false when the list is empty and
  *    the value is not nullable.
  * 2. Removes literal repetitions.
- * 3. Replaces [[In (value, seq[Literal])]] with optimized version
+ * 3. Replaces [[In (values, seq[Literal])]] with optimized version
  *    [[InSet (value, HashSet[Literal])]] which is much faster.
  */
 object OptimizeIn extends Rule[LogicalPlan] {
   def apply(plan: LogicalPlan): LogicalPlan = plan transform {
     case q: LogicalPlan => q transformExpressionsDown {
-      case In(v, list) if list.isEmpty =>
+      case i @ In(_, list) if list.isEmpty =>
         // When v is not nullable, the following expression will be optimized
         // to FalseLiteral which is tested in OptimizeInSuite.scala
-        If(IsNotNull(v), FalseLiteral, Literal(null, BooleanType))
-      case expr @ In(v, list) if expr.inSetConvertible =>
+        If(IsNotNull(i.value), FalseLiteral, Literal(null, BooleanType))
+      case expr @ In(_, list) if expr.inSetConvertible =>
         val newList = ExpressionSet(list).toSeq
         if (newList.length == 1
           // TODO: `EqualTo` for structural types are not working. Until SPARK-24443 is addressed,
           // TODO: we exclude them in this rule.
-          && !v.isInstanceOf[CreateNamedStructLike]
+          && !expr.value.isInstanceOf[CreateNamedStructLike]
```

Comment:
```
  @transient protected lazy val isMultiValued = values.length > 1
  @transient lazy val value: Expression = if (isMultiValued) {
    CreateNamedStruct(values.zipWithIndex.flatMap {
      case (v: NamedExpression, _) => Seq(Literal(v.name), v)
      case (v, idx) => Seq(Literal(s"_$idx"), v)
    })
  } else {
    values.head
  }
}
```

According to the implementation, `expr.value.isInstanceOf[CreateNamedStructLike]` means `expr.values.length > 1`, right?

Reply:
yes, rigth

Reply:
shall we use `expr.values.length == 1` here to make it more clear?

Reply:
not really, because `expr.value.isInstanceOf[CreateNamedStructLike]` means:

 - either `expr.values.length == 1`;
 - or `expr.values.head.isInstanceOf[CreateNamedStructLike]`;

Basically there are 2 cases: one where we have several attributes in the value before IN; the other when there is a single value before IN but the value is a struct. `expr.value.isInstanceOf[CreateNamedStructLike]` catches both. I can add a comment explaining these 2 cases if you think is needed.

Reply:
IIUC, you mean
`expr.values.length > 1` => `expr.value.isInstanceOf[CreateNamedStructLike]`
but `expr.value.isInstanceOf[CreateNamedStructLike]` can't => `expr.values.length > 1`

Can you give an example?

Based on my understanding, the code here is trying to optimize a case when it's not a multi-value in and the list has only one element.

Reply:
yes, I mean that. An example is:
```
create table t1 as select struct('a', 1, 'b', '2') as a1;
select 1 from (select struct('a', 1, 'b', '2') as a1) t1 where a1 in ((...), ...);
```

Reply:
for your case, it's not `CreateNamedStructLike`, but just a struct type column?

Reply:
no, because of optimizations, it is a `CreateNamedStructLike`

Reply:
well, I think for this case we should optimize it.

Anyway it follows the previous behavior, we can change it later.